File: Web Services and Middleware Technologies/Quicksort for matrix lines/Java/src/MatrixQuickSort.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.*;

public class MatrixQuickSort {

    private static final List<List<String>> matrix = new ArrayList<>();

    private static void readFromFile(String filename) {

        int rows = 0;
        int columns = 0;
        List<String> strings = new ArrayList<>();
        try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
            String line;
            while ((line = br.readLine()) != null) {
                String[] s = line.split("[,]");
                if (s.length != 3) {
                    System.out.println("Incorrect line");
                    continue;
                }
                if (Integer.parseInt(s[0]) > rows)
                    rows = Integer.parseInt(s[0]);
                if (Integer.parseInt(s[1]) > columns)
                    columns = Integer.parseInt(s[1]);
                strings.add(line);
            }
        } catch (IOException e) {
            System.out.println("Read from file error!\n");
        }
        rows++;
        columns++;
        for (int i = 0; i < rows; i++) {
            List<String> line = new ArrayList<>(Collections.nCopies(columns, null));
            matrix.add(line);
        }
        for (String s : strings) {
            String[] strs = s.split("[,]");
            int l = Integer.parseInt(strs[0]);
            int c = Integer.parseInt(strs[1]);
            matrix.get(l).set(c, strs[2]);
        }
    }

    private static void printMatrix() {

        for (List<String> strings : matrix) {
            for (String str : strings) System.out.print(str + "  ");
            System.out.println();
        }
    }

    private static void randomizedQuickSort3(int up, int down) {

        if (up >= down)
            return;

        Random random = new Random();
        int k = up + Math.abs(random.nextInt()) % (down - up + 1);
        List<String> temp = matrix.get(k);
        matrix.set(k, matrix.get(up));
        matrix.set(up, temp);
        List<Integer> positions = partition3(up, down);

        randomizedQuickSort3(up, positions.get(0) - 1);
        randomizedQuickSort3(positions.get(1) + 1, down);
    }

    private static List<Integer> partition3(int up, int down) {

        List<String> x = matrix.get(up);
        int j = up;
        for (int i = up + 1; i <= down; i++) {
            Boolean comp = compareLines(matrix.get(i), x);
            if (comp == null) continue;
            if (comp) {
                List<String> temp = matrix.get(j);
                matrix.set(j, matrix.get(i));
                matrix.set(i, temp);
                j++;
            } else {
                List<String> temp = matrix.get(i);
                matrix.set(i, matrix.get(down));
                matrix.set(down, temp);
                down--;
                i--;
            }
        }
        return Arrays.asList(j, down);
    }

    private static Boolean compareLines(List<String> l1, List<String> l2) {

        for (int i = 0; i < l1.size(); i++) {
            if (l1.get(i).equals(l2.get(i)))
                continue;
            if (l1.get(i) == null)
                return true;
            if (l2.get(i) == null)
                return false;
            return l1.get(i).compareTo(l2.get(i)) < 0;
        }
        return null;
    }

    public static void main(String[] args) {

        Scanner scanner = new Scanner(System.in);
        String filename = "..\\files\\";
        if (args.length == 0) {
            System.out.println("Read the file name: ");
            filename += scanner.next();
        } else {
            filename += args[0];
        }
        System.out.println(filename);
        readFromFile(filename);
        System.out.println("\nInitial matrix");
        printMatrix();
        randomizedQuickSort3(0, matrix.size() - 1);
        System.out.println("\nSorted matrix");
        printMatrix();
    }
}

File: Web Services and Middleware Technologies/Quicksort for matrix lines/CSharp/MatrixQuicksort.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Collections.Specialized;
using System.IO;
using System.Linq;
using static System.Int32;

namespace CSharp
{
    internal class MatrixQuickSort
    {
        private static List<List<string>> matrix = new List<List<string>>();

        private static void ReadFromFile(string filename)
        {
            var rows = 0;
            var columns = 0;
            var strings = new List<string>();
            using (TextReader tr = File.OpenText(filename))
            {
                string line;
                while ((line = tr.ReadLine()) != null)
                {
                    var s = line.Split(',');
                    if (s.Length != 3)
                    {
                        Console.WriteLine("Incorrect line");
                        continue;
                    }

                    if (Parse(s[0]) > rows)
                        rows = Parse(s[0]);
                    if (Parse(s[1]) > columns)
                        columns = Parse(s[1]);
                    strings.Add(line);
                }
            }

            rows++;
            columns++;
            for (var i = 0; i < rows; i++)
            {
                var line = new List<string>(columns);
                line.AddRange(Enumerable.Repeat<string>(null, columns));
                matrix.Add(line);
            }

            foreach (var s in strings)
            {
                var strs = s.Split(',');
                var l = Parse(strs[0]);
                var c = Parse(strs[1]);
                matrix[l][c] = strs[2];
            }
        }

        private static void PrintMatrix()
        {
            foreach (var strings in matrix)
            {
                foreach (var str in strings)
                    Console.Write(str + " ");
                Console.WriteLine();
            }
        }

        private static void RandomizedQuickSort3(int up, int down)
        {
            if (up >= down)
                return;

            var random = new Random();
            var k = up + random.Next() % (down - up + 1);
            var temp = matrix[k];
            matrix[k] = matrix[up];
            matrix[up] = temp;
            var positions = Partition3(up, down);

            RandomizedQuickSort3(up, positions[0] - 1);
            RandomizedQuickSort3(positions[1] + 1, down);
        }

        private static List<int> Partition3(int up, int down)
        {
            var x = matrix[up];
            var j = up;
            for (var i = up + 1; i <= down; i++)
            {
                var comp = CompareLines(matrix[i], x);
                if (comp == null) continue;
                if (comp.Equals(true))
                {
                    var temp = matrix[j];
                    matrix[j] = matrix[i];
                    matrix[i] = temp;
                    j++;
                }
                else
                {
                    var temp = matrix[i];
                    matrix[i] = matrix[down];
                    matrix[down] = temp;
                    down--;
                    i--;
                }
            }

            return new List<int>(new[] {j, down});
        }

        private static bool? CompareLines(List<string> l1, IReadOnlyList<string> l2)
        {
            for (var i = 0; i < l1.Capacity; i++)
            {
                if (l1[i].Equals(l2[i]))
                    continue;
                if (l1[i] == null)
                    return true;
                if (l2[i] == null)
                    return false;
                return string.Compare(l1[i], l2[i], StringComparison.Ordinal) < 0;
            }

            return null;
        }

        public static void Main(string[] args)
        {
            var filename = "files\\";
            if (args.Length == 0)
            {
                Console.WriteLine("Read the file name: ");
                filename += Console.ReadLine();
            }
            else
            {
                filename += args[0];
            }
            ReadFromFile(filename);
            Console.WriteLine("\nInitial matrix");
            PrintMatrix();
            RandomizedQuickSort3(0, matrix.Capacity - 1);
            Console.WriteLine("\nSorted matrix");
            PrintMatrix();
        }
    }
}